File: neuron_db/secure.py
from __future__ import annotations
import os, hmac, hashlib, base64, json, sqlite3, threading, time
from typing import Optional
from .neuron import _content, _stem
try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    _HAS_AESGCM = True
except Exception:
    _HAS_AESGCM = False
def _hkdf(key, salt, info, n=32):
    prk=hmac.new(salt,key,hashlib.sha256).digest(); out=b""; t=b""; i=1
    while len(out)<n:
        t=hmac.new(prk,t+info+bytes([i]),hashlib.sha256).digest(); out+=t; i+=1
    return out[:n]
def aead_encrypt(key, plaintext, aad=b""):
    nonce=os.urandom(12)
    if _HAS_AESGCM:
        return b"\x01"+nonce+AESGCM(_hkdf(key,b"neuron-aesgcm",b"v1",32)).encrypt(nonce,plaintext,aad)
    ek=_hkdf(key,nonce,b"enc"); mk=_hkdf(key,nonce,b"mac"); ks=b""; c=0
    while len(ks)<len(plaintext): ks+=hmac.new(ek,nonce+c.to_bytes(8,"big"),hashlib.sha256).digest(); c+=1
    ct=bytes(a^b for a,b in zip(plaintext,ks))
    tag=hmac.new(mk,b"\x00"+nonce+aad+ct,hashlib.sha256).digest()[:16]
    return b"\x00"+nonce+tag+ct
def aead_decrypt(key, blob, aad=b""):
    ver,nonce=blob[:1],blob[1:13]
    if ver==b"\x01":
        return AESGCM(_hkdf(key,b"neuron-aesgcm",b"v1",32)).decrypt(nonce,blob[13:],aad)
    tag,ct=blob[13:29],blob[29:]; mk=_hkdf(key,nonce,b"mac")
    if not hmac.compare_digest(tag,hmac.new(mk,b"\x00"+nonce+aad+ct,hashlib.sha256).digest()[:16]):
        raise ValueError("authentication failed: wrong key or tampered data")
    ek=_hkdf(key,nonce,b"enc"); ks=b""; c=0
    while len(ks)<len(ct): ks+=hmac.new(ek,nonce+c.to_bytes(8,"big"),hashlib.sha256).digest(); c+=1
    return bytes(a^b for a,b in zip(ct,ks))
def derive_key(secret, neuron_id):
    return _hkdf(secret.encode(),b"neuron-db/"+neuron_id.encode(),b"key",32)
class SecureNeuron:
    def __init__(self, key, entries=None):
        self._key=key; self._idx_key=_hkdf(key,b"neuron-index",b"v1",32); self.entries=entries or []
    def _keyed(self, stem):
        return base64.b64encode(hmac.new(self._idx_key,stem.encode(),hashlib.sha256).digest()[:8]).decode()
    def _stems(self, phrase): return {self._keyed(s) for s in _stem(_content(phrase))}
    def put(self, key_phrase, value):
        idx=sorted(self._stems(key_phrase))
        if not idx: raise ValueError("key phrase has no indexable content")
        ct=base64.b64encode(aead_encrypt(self._key,value.encode())).decode()
        self.entries.append({"x":idx,"c":ct})
    def get(self, query, min_cover=0.5):
        q=self._stems(query)
        if not q: return None
        best,bk=None,(-1.0,-1)
        for i,e in enumerate(self.entries):
            xs=set(e["x"]); ov=len(q&xs)
            if not ov: continue
            cover=ov/len(xs)
            if cover<min_cover: continue
            sc=(cover,i)
            if sc>bk: bk,best=sc,e
        if best is None: return None
        try: return aead_decrypt(self._key,base64.b64decode(best["c"])).decode()
        except Exception: return None
    def dump(self): return json.dumps(self.entries,separators=(",",":"))
    @classmethod
    def load(cls, key, blob): return cls(key,json.loads(blob or "[]"))
    @property
    def count(self): return len(self.entries)
class SecureNeuronDB:
    SCHEMA="CREATE TABLE IF NOT EXISTS secure (id TEXT PRIMARY KEY, blob TEXT NOT NULL, updated INTEGER NOT NULL);"
    def __init__(self, path="secure.db"):
        self.conn=sqlite3.connect(path,check_same_thread=False); self._lock=threading.Lock()
        self.conn.execute(self.SCHEMA); self.conn.commit()
    def _blob(self, nid):
        row=self.conn.execute("SELECT blob FROM secure WHERE id=?",(nid,)).fetchone(); return row[0] if row else "[]"
    def _write(self, nid, blob):
        self.conn.execute("INSERT INTO secure(id,blob,updated) VALUES(?,?,?) ON CONFLICT(id) DO UPDATE SET blob=excluded.blob,updated=excluded.updated",(nid,blob,int(time.time()*1000)))
        self.conn.commit()
    def put(self, nid, secret, key_phrase, value):
        with self._lock:
            k=derive_key(secret,nid); n=SecureNeuron.load(k,self._blob(nid)); n.put(key_phrase,value); self._write(nid,n.dump())
    def get(self, nid, secret, query):
        with self._lock:
            k=derive_key(secret,nid); n=SecureNeuron.load(k,self._blob(nid)); return n.get(query)
    def close(self): self.conn.close()

File: rust/neuron-core/src/secure.rs
//! SecureNeuronDB: encrypted neurons. Values are AES-256-GCM ciphertext; the index is a
//! keyed hash of cue stems; the per-neuron secret is supplied per call and never stored, so
//! a stolen db file is opaque. Feature-gated behind `secure`. Faithful port of
//! neuron_db/secure.py (HKDF-SHA256 key schedule; Rust-native blob format).
use aes_gcm::{Aes256Gcm, Key, Nonce};
use aes_gcm::aead::{Aead, KeyInit, Payload};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use base64::{Engine, engine::general_purpose::STANDARD as B64};
use std::collections::BTreeSet;
use crate::{content, stems_s};

type HmacSha256 = Hmac<Sha256>;
fn mac(key: &[u8], msg: &[u8]) -> Vec<u8> {
    let mut m = <HmacSha256 as Mac>::new_from_slice(key).expect("hmac key");
    m.update(msg); m.finalize().into_bytes().to_vec()
}
/// HKDF-SHA256 (extract+expand), matching the Python _hkdf.
fn hkdf(key: &[u8], salt: &[u8], info: &[u8], n: usize) -> Vec<u8> {
    let prk = mac(salt, key);
    let mut out = Vec::new(); let mut t: Vec<u8> = Vec::new(); let mut i = 1u8;
    while out.len() < n {
        let mut msg = t.clone(); msg.extend_from_slice(info); msg.push(i);
        t = mac(&prk, &msg); out.extend_from_slice(&t); i = i.wrapping_add(1);
    }
    out.truncate(n); out
}
pub fn derive_key(secret: &str, nid: &str) -> Vec<u8> {
    let mut salt = b"neuron-db/".to_vec(); salt.extend_from_slice(nid.as_bytes());
    hkdf(secret.as_bytes(), &salt, b"key", 32)
}
// Additional authenticated data bound into every GCM tag, so a ciphertext can't be silently
// replayed under a different format version. The blob's leading byte is the format version: 1 =
// legacy (empty AAD, still readable), 2 = AAD-bound (what we write now).
const AAD_V2: &[u8] = b"neuron-secure-v2";
fn aead_encrypt(key: &[u8], pt: &[u8]) -> Vec<u8> {
    let mut nonce = [0u8; 12]; getrandom::getrandom(&mut nonce).expect("rng");
    let k = hkdf(key, b"neuron-aesgcm", b"v1", 32);
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&k));
    let ct = cipher.encrypt(Nonce::from_slice(&nonce), Payload { msg: pt, aad: AAD_V2 }).expect("encrypt");
    let mut out = vec![2u8]; out.extend_from_slice(&nonce); out.extend_from_slice(&ct); out
}
fn aead_decrypt(key: &[u8], blob: &[u8]) -> Option<Vec<u8>> {
    if blob.len() < 13 { return None; }
    let aad: &[u8] = match blob[0] { 1 => b"", 2 => AAD_V2, _ => return None };  // read legacy + AAD-bound
    let k = hkdf(key, b"neuron-aesgcm", b"v1", 32);
    let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&k));
    cipher.decrypt(Nonce::from_slice(&blob[1..13]), Payload { msg: &blob[13..], aad }).ok()
}

pub struct SecureNeuron { key: Vec<u8>, idx_key: Vec<u8>, entries: Vec<(Vec<String>, String)> }
impl SecureNeuron {
    pub fn new(key: Vec<u8>) -> Self {
        let idx_key = hkdf(&key, b"neuron-index", b"v1", 32);
        SecureNeuron { key, idx_key, entries: Vec::new() }
    }
    fn keyed(&self, stem: &str) -> String { B64.encode(&mac(&self.idx_key, stem.as_bytes())[..8]) }
    fn stems(&self, phrase: &str) -> BTreeSet<String> {
        stems_s(&content(phrase)).iter().map(|s| self.keyed(s)).collect()
    }
    pub fn put(&mut self, key_phrase: &str, value: &str) -> Result<(), &'static str> {
        let idx: Vec<String> = self.stems(key_phrase).into_iter().collect(); // BTreeSet -> sorted
        if idx.is_empty() { return Err("key phrase has no indexable content"); }
        let ct = B64.encode(aead_encrypt(&self.key, value.as_bytes()));
        self.entries.push((idx, ct)); Ok(())
    }
    pub fn get(&self, query: &str, min_cover: f64) -> Option<String> {
        let q = self.stems(query); if q.is_empty() { return None; }
        let mut best: Option<&(Vec<String>, String)> = None; let mut bk = (-1f64, -1i64);
        for (i, e) in self.entries.iter().enumerate() {
            let xs: BTreeSet<&String> = e.0.iter().collect();
            let ov = q.iter().filter(|s| xs.contains(s)).count();
            if ov == 0 { continue; }
            let cover = ov as f64 / e.0.len() as f64;
            if cover < min_cover { continue; }
            if cover > bk.0 || (cover == bk.0 && i as i64 > bk.1) { bk = (cover, i as i64); best = Some(e); }
        }
        let e = best?;
        let raw = B64.decode(&e.1).ok()?;
        let pt = aead_decrypt(&self.key, &raw)?;
        String::from_utf8(pt).ok()
    }
    pub fn dump(&self) -> String {
        self.entries.iter().map(|(idx, ct)| format!("{}|{}", idx.join(","), ct)).collect::<Vec<_>>().join("\n")
    }
    pub fn load(key: Vec<u8>, blob: &str) -> Self {
        let mut n = SecureNeuron::new(key);
        for line in blob.split('\n') {
            if line.is_empty() { continue; }
            if let Some((ix, ct)) = line.split_once('|') {
                let idx = ix.split(',').filter(|s| !s.is_empty()).map(|s| s.to_string()).collect();
                n.entries.push((idx, ct.to_string()));
            }
        }
        n
    }
    pub fn count(&self) -> usize { self.entries.len() }
}

use rusqlite::{params, Connection};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};
const SCHEMA: &str = "CREATE TABLE IF NOT EXISTS secure (id TEXT PRIMARY KEY, blob TEXT NOT NULL, updated INTEGER NOT NULL);";
pub struct SecureNeuronDB { conn: Mutex<Connection> }
impl SecureNeuronDB {
    pub fn open(path: &str) -> Self {
        let conn = Connection::open(path).expect("open");
        conn.execute(SCHEMA, []).expect("schema");
        SecureNeuronDB { conn: Mutex::new(conn) }
    }
    fn blob(&self, conn: &Connection, nid: &str) -> String {
        conn.query_row("SELECT blob FROM secure WHERE id=?1", params![nid], |r| r.get::<_, String>(0)).unwrap_or_else(|_| String::new())
    }
    fn write(&self, conn: &Connection, nid: &str, blob: &str) {
        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as i64;
        conn.execute("INSERT INTO secure(id,blob,updated) VALUES(?1,?2,?3) ON CONFLICT(id) DO UPDATE SET blob=excluded.blob,updated=excluded.updated", params![nid, blob, now]).expect("write");
    }
    pub fn put(&self, nid: &str, secret: &str, key_phrase: &str, value: &str) -> Result<(), &'static str> {
        let conn = self.conn.lock().unwrap();
        let k = derive_key(secret, nid);
        let mut n = SecureNeuron::load(k, &self.blob(&conn, nid));
        n.put(key_phrase, value)?; self.write(&conn, nid, &n.dump()); Ok(())
    }
    pub fn get(&self, nid: &str, secret: &str, query: &str) -> Option<String> {
        let conn = self.conn.lock().unwrap();
        let k = derive_key(secret, nid);
        let n = SecureNeuron::load(k, &self.blob(&conn, nid));
        n.get(query, 0.5)
    }
}
